runtimes/starlarkrt/runtime/builtins.go

```diff
@@ -20,17 +22,69 @@ var builtins = map[string]func(thread *starlark.Thread, fn *starlark.Builtin, ar
 
 		value, err := starlark.Call(thread, fn, args[1:], kwargs)
 		if err != nil {
+			if perr, ok := sdktypes.FromError(err); ok {
+				v := perr.Value()
+				vctx := values.Context{}
+				if slv, cerr := vctx.ToStarlarkValue(v); cerr == nil {
+					return starlark.Tuple{starlark.None, slv}, nil
+				}
+			}
+
 			return starlark.Tuple{starlark.None, starlark.String(err.Error())}, nil
 		}
 
 		return starlark.Tuple{value, starlark.None}, nil
 	},
 	"fail": func(thread *starlark.Thread, bi *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
-		var msg string
-		if err := starlark.UnpackArgs(bi.Name(), args, kwargs, "msg", &msg); err != nil {
-			return nil, err
+		if len(args) > 0 && len(kwargs) > 0 {
+			return nil, fmt.Errorf("fail: cannot specify both positional and keyword arguments")
 		}
-		return nil, errors.New(msg)
+
+		vctx := values.Context{}
+
+		if len(args) == 1 {
+			v, err := vctx.FromStarlarkValue(args[0])
+			if err != nil {
+				return nil, fmt.Errorf("cannot convert value from starlark: %w", err)
+			}
+			return nil, sdktypes.NewProgramError(v, nil, nil).ToError()
+		} else if len(args) > 1 {
+			vs, err := kittehs.TransformError(args, vctx.FromStarlarkValue)
+			if err != nil {
+				return nil, fmt.Errorf("cannot convert values from starlark: %w", err)
+			}
+
+			v := sdktypes.NewListValue(vs)
+			return nil, sdktypes.NewProgramError(v, nil, nil).ToError()
+		}
+
+		if len(kwargs) > 0 {
+			vs, err := kittehs.ListToMapError(kwargs, func(t starlark.Tuple) (string, sdktypes.Value, error) {
+				if len(t) != 2 {
+					return "", sdktypes.InvalidValue, fmt.Errorf("expected 2 elements, got %d", len(t))
+				}
+
+				k, ok := t[0].(starlark.String)
+				if !ok {
+					return "", sdktypes.InvalidValue, fmt.Errorf("expected string, got %T", t[0])
+				}
+
+				v, err := vctx.FromStarlarkValue(t[1])
+				if err != nil {
+					return "", sdktypes.InvalidValue, fmt.Errorf("cannot convert value from starlark: %w", err)
+				}
+
+				return k.GoString(), v, nil
+			})
+			if err != nil {
+				return nil, fmt.Errorf("cannot convert values from starlark: %w", err)
+			}
+
+			v := sdktypes.NewDictValueFromStringMap(vs)
+			return nil, sdktypes.NewProgramError(v, nil, nil).ToError()
+		}
+
+		return nil, errors.New("user triggered error")
```

Comment:
move this to the top after the check that not both kwargs and args len > 0
and then you can remove this last condition as it is implicit
again less indentation

pseudo code
so if both > 0 fail
if both == 0 fail

if args == 1 return do something
if args > 1 return do something 2
do something 3 // implicit len(args)=0 and kwargs len > 0

Reply:
rearranged to be more clear